### PR TITLE
Use Get instead of List for DNS zones and records

### PR DESCRIPTION
**Pull request by kzwang**
**What this PR does / why we need it**:
Currently, federation DNS provider uses List to find DNS zone and record, this is inefficient if there are lots of zones or lots of records in the zone. This PR adds a Get method to DNS provider interface for zones and record sets.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37506)

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

File: federation/pkg/dnsprovider/providers/aws/route53/route53_test.go [hunk 1]
```diff
@@ -146,14 +154,29 @@ func TestZonesList(t *testing.T) {
...
 
-/* TestZonesID verifies that the id of the zone is returned with the prefix removed */
-func TestZonesID(t *testing.T) {
+/* TestFindZoneByID verifies that getting zone by id succeeds */
+func TestFindZoneByID(t *testing.T) {
 	zone := firstZone(t)
+	z := zones(t)
+	zoneById, _ := z.Get("", zone.ID())
+	if zoneById.ID() != zone.ID() {
+		t.Fatalf("Unexpected zone id: %q", zone.ID())
```
Comment (nikhiljindal) [inline federation/pkg/dnsprovider/providers/aws/route53/route53_test.go]: Also output the expected zone Id. Here and everywhere

File: federation/pkg/dnsprovider/providers/aws/route53/rrsets.go [hunk 2]
```diff
@@ -49,18 +51,31 @@ func (rrsets ResourceRecordSets) List() ([]dnsprovider.ResourceRecordSet, error)
...
-	for _, rrset := range rrsetList {
-		if rrset.Name() == name {
-			newRrset = rrset
-			break
-		}
+	if len(list) == 0 {
+		return nil, nil
+	}
+	if len(list) > 1 {
+		return nil, fmt.Errorf("DNS name %s is ambiguous.", name)
```
Comment (nikhiljindal) [inline federation/pkg/dnsprovider/providers/aws/route53/rrsets.go]: This is a behavior change. Earlier, we returned the first record in case of more than one records. With this change we will return an error.
cc @quinton-hoole to confirm if this looks good to him.

File: federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go [hunk 3]
```diff
@@ -108,9 +115,28 @@ func (r *Route53APIStub) ListHostedZonesPages(input *route53.ListHostedZonesInpu
...
+	if !ok {
+		return nil, fmt.Errorf("Error getting hosted DNS zone: %s does not exists", id)
+	}
+	return &route53.GetHostedZoneOutput{HostedZone: zone}, nil
+}
+
 func (r *Route53APIStub) CreateHostedZone(input *route53.CreateHostedZoneInput) (*route53.CreateHostedZoneOutput, error) {
 	name := aws.StringValue(input.Name)
-	id := "/hostedzone/" + name
+	id := "/hostedzone/" + string(uuid.NewUUID())
```
Comment (nikhiljindal) [inline federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go]: Why this change? Are both correct or was the previous code wrong?

File: federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zone_create_call.go [hunk 4]
```diff
@@ -37,16 +37,16 @@ func (call ManagedZonesCreateCall) Do(opts ...googleapi.CallOption) (interfaces.
 	if call.Error != nil {
 		return nil, *call.Error
 	}
-	if call.Service.Impl[call.Project][call.ManagedZone.DnsName()] != nil {
+	if call.Service.Impl[call.Project][call.ManagedZone.Name()] != nil {
 		return nil, fmt.Errorf("Error - attempt to create duplicate zone %s in project %s.",
-			call.ManagedZone.DnsName(), call.Project)
+			call.ManagedZone.Name(), call.Project)
```
Comment (nikhiljindal) [inline federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zone_create_call.go]: Same. Are both correct or was the previous code wrong?

**Comment by nikhiljindal** [comment]
Thanks for the PR @kzwang 
Can you please confirm that you ran the federated service e2e test and it passes with this change?

cc @kubernetes/sig-federation 
Also cc @quinton-hoole who wrote most of this code and @justinsb and @shashidharatd who are familiar with route 53 DNS and CoreDNS implementations respectively.

File: federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go [hunk 5]
```diff
@@ -61,7 +65,10 @@ func (r *Route53APIStub) ListResourceRecordSetsPages(input *route53.ListResource
 	} else {
 		for _, rrsets := range r.recordSets[*input.HostedZoneId] {
 			for _, rrset := range rrsets {
-				output.ResourceRecordSets = append(output.ResourceRecordSets, rrset)
+				if input.StartRecordName == nil ||
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go]: Please add as comment to explain what you're doing here.  It's not obvious.

**Comment by nikhiljindal** [comment]
Ping @kzwang

File: federation/pkg/dnsprovider/providers/aws/route53/rrsets.go (hunk 2 above)
Comment (kzwang) [inline federation/pkg/dnsprovider/providers/aws/route53/rrsets.go]: What I'm thinking is if the DNS record is managed by this code, then it should never have two records with the same name. If it has two records with the same name, then that record has been modified by the user or some other application. I think it's better to log that as an error instead of silently modify the record.

File: federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go (hunk 3 above)
Comment (kzwang) [inline federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go]: Both are correct. I changed this because in AWS the id is not the same as the DNS name. I don't want test pass because the DNS name happens to be same as id

**Comment by nikhiljindal** [comment]
I see that you rebased @kzwang 
Can you please also reply to comments from @quinton-hoole

File: federation/pkg/dnsprovider/providers/aws/route53/rrsets.go (hunk 2 above)
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/aws/route53/rrsets.go]: @kzwang That's a reasonable argument, except that in general in we adopt an "anti-entropy" or "reconciliation" approach, where controllers will adopt and update underlying entities when they exist and are inconsistent.  We do that in many other cases, particularly in Federation.  I think that we should keep consistent with that approach here.

**Comment by kzwang** [comment]
@quinton-hoole I've changed the behavior back to return first matching RRS

**Comment by kzwang** [comment]
@k8s-bot gce etcd3 e2e test this

File: federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zone_create_call.go (hunk 4 above)
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zone_create_call.go]: In the original implementation ManagedZone.Name() and ManagedZone.DnsName() were set to be equal, so both work.  But I see that e.g. [ManagedZonesDeleteCall.Do()](https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zones_delete_call.go#L44) uses ManagedZone.Name() as the key, so this change improves code consistency.
From https://godoc.org/google.golang.org/api/dns/v1#ManagedZone 

```
// DnsName: The DNS name of this managed zone, for instance
// "example.com.".

// Name: User assigned name for this resource. Must be unique within the
// project. The name must be 1-63 characters long, must begin with a
// letter, end with a letter or digit, and only contain lowercase
// letters, digits or dashes.

File: federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/rrsets_list_call.go [hunk 6]
```diff
@@ -19,19 +19,30 @@ package stubs
...
-	Type_     string
+	Response_   *ResourceRecordSetsListResponse
+	Err_        error
+	Name_       string
+	Type_       string
+	MaxResults_ int64
 }
 
 func (call *ResourceRecordSetsListCall) Do(opts ...googleapi.CallOption) (interfaces.ResourceRecordSetsListResponse, error) {
+	if len(call.Name_) > 0 {
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/google/clouddns/internal/stubs/rrsets_list_call.go]: It's not at all clear why all this additional logic is required here.  Please add a comment to explain, or remove it.

File: federation/pkg/dnsprovider/providers/google/clouddns/rrsets.go [hunk 7]
```diff
@@ -43,18 +44,20 @@ func (rrsets ResourceRecordSets) List() ([]dnsprovider.ResourceRecordSet, error)
...
 	if err != nil {
 		return nil, err
 	}
-	for _, rrset := range rrsetList {
-		if rrset.Name() == name {
-			newRrset = rrset
-			break
+	list := make([]dnsprovider.ResourceRecordSet, 0)
+	for _, rrset := range response.Rrsets() {
+		if strings.TrimSuffix(name, ".") == strings.TrimSuffix(rrset.Name(), ".") {
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/google/clouddns/rrsets.go]: This trim logic occurs in several different places, and should be factored out into a common utility function, with a comment to explain why it's necessary.

File: federation/pkg/dnsprovider/providers/google/clouddns/zones.go [hunk 8]
```diff
@@ -42,6 +44,36 @@ func (zones Zones) List() ([]dnsprovider.Zone, error) {
 	return zoneList, nil
 }
 
+func (zones Zones) Get(dnsZoneName string, dnsZoneID string) (dnsprovider.Zone, error) {
+	if len(dnsZoneID) > 0 {
```
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/providers/google/clouddns/zones.go]: This logic seems complex and unnecessary.  For a given cloud provider, is it not the case that one of dnsZoneName or dnsZoneId should always be empty, and the other always non-empty?  Also, same comment above regarding the unclear and duplicated trim logic.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by quinton-hoole** [comment]
Thanks for the update @kzwang .
Please see my additional comments.

And to repeat @nikhiljindal 's earlier question:

Can you please confirm that you ran the federated service e2e test and it passes with this change?
